presto-raptorx/src/main/java/com/facebook/presto/raptorx/RaptorConnector.java

```diff
@@ -0,0 +1,179 @@
+    private final List<PropertyMetadata<?>> sessionProperties;
+    private final List<PropertyMetadata<?>> tableProperties;
+    private final Set<Procedure> procedures;
+    private final Set<SystemTable> systemTables;
+
+    @Inject
+    public RaptorConnector(
+            LifeCycleManager lifeCycleManager,
+            TransactionManager transactionManager,
+            NodeSupplier nodeSupplier,
+            BucketManager bucketManager,
+            TransactionWriter transactionWriter,
+            ConnectorSplitManager splitManager,
+            ConnectorPageSourceProvider pageSourceProvider,
+            ConnectorPageSinkProvider pageSinkProvider,
+            ConnectorNodePartitioningProvider nodePartitioningProvider,
+            RaptorSessionProperties sessionProperties,
+            RaptorTableProperties tableProperties,
+            Set<Procedure> procedures,
+            Set<SystemTable> systemTables)
+    {
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.nodeSupplier = requireNonNull(nodeSupplier, "nodeSupplier is null");
+        this.bucketManager = requireNonNull(bucketManager, "bucketManager is null");
+        this.transactionWriter = requireNonNull(transactionWriter, "transactionWriter is null");
+        this.splitManager = requireNonNull(splitManager, "splitManager is null");
+        this.pageSourceProvider = requireNonNull(pageSourceProvider, "pageSourceProvider is null");
+        this.pageSinkProvider = requireNonNull(pageSinkProvider, "pageSinkProvider is null");
+        this.nodePartitioningProvider = requireNonNull(nodePartitioningProvider, "nodePartitioningProvider is null");
+        this.sessionProperties = requireNonNull(sessionProperties, "sessionProperties is null").getSessionProperties();
+        this.tableProperties = requireNonNull(tableProperties, "tableProperties is null").getTableProperties();
+        this.procedures = ImmutableSet.copyOf(requireNonNull(procedures, "procedures is null"));
+        this.systemTables = ImmutableSet.copyOf(requireNonNull(systemTables, "systemTables is null"));
+    }
+
+    @Override
+    public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel, boolean readOnly)
+    {
+        checkConnectorSupports(REPEATABLE_READ, isolationLevel);
+        return transactionManager.create();
+    }
+
+    @Override
+    public void commit(ConnectorTransactionHandle transactionHandle)
+    {
+        Transaction transaction = transactionManager.remove(transactionHandle);
+        transactionWriter.write(transaction.getActions(), transaction.getRegisteredTransactionId());
+    }
+
+    @Override
+    public void rollback(ConnectorTransactionHandle transactionHandle)
+    {
+        transactionManager.remove(transactionHandle);
```

Comment:
There is a low chance unsafe state after line 111 and before line 112. 
Once transactionManager.remove, CommitCleaner can do some work, then transactionWriter.recover may not be able to rollback

Reply:
I removed the recover call as it's not needed -- the engine will not call both commit and abort. The idea was to optimistically recover, rather than waiting for the next commit. However, it won't be used.

Reply:
Yes. Solid.